### ISSUE TITLE
Logit Discrepancy Between Candle and PyTorch When Using XLM-RoBERTa Model

### ISSUE THREAD

**Issue by jpe90** [huggingface/candle#2967]
When running the same XLM-RoBERTa model (`s-nlp/xlmr_formality_classifier` - [HF](https://huggingface.co/s-nlp/xlmr_formality_classifier) ) in both Candle and PyTorch, I'm observing significant differences in the logits produced by the model's classification head for identical inputs. Is this expected behavior? See [this repository](https://github.com/jpe90/candle-pytorch-parity-testing/tree/master/xlm-roberta-finetuned) for a reproduction.

## Environment/Setup

- Model: `s-nlp/xlmr_formality_classifier` 
- Candle version: 0.9.1
- Model SHA256: `66037d963856d6d001f3109d2b3cf95c76bce677947e66f426299c89bc1b58e7`
- OS: macOS

## Observed Behavior

Given identical inputs, the logits produced by Candle and PyTorch differ significantly:

**Candle logits:**
```
[[2.0820313, -1.7548828], [0.7783203, -0.5629883], [1.2871094, -1.0039063], [2.1601563, -1.9277344]]
```

**PyTorch logits:**
```
[[ 2.6433, -2.3445],
 [ 1.0379, -0.9621],
 [ 1.4154, -1.2704],
 [ 3.4423, -3.1726]]
```

## Expected Behavior

I would expect the logits to be extremely close (within floating-point precision differences) when running the same model with identical inputs across different frameworks.

## Steps to Reproduce

1. Clone the repository: https://github.com/jpe90/candle-pytorch-parity-testing
2. Run the PyTorch implementation in `/xlm-roberta-finetuned/pytorch/main.py`
3. Run the Candle implementation in `/xlm-roberta-finetuned/candle/src/main.rs`
4. Compare the logits produced by both implementations

## Additional Context

- The tokenization appears to be identical between both implementations (identical token IDs)
- I checked and made sure model checksums match at runtime
- Config seems to match ([see here](https://github.com/jpe90/candle-pytorch-parity-testing/blob/master/xlm-roberta-finetuned/troubleshooting.md))

## Questions

1. Should I expect identical (or very close) logits between PyTorch and Candle implementations?
2. If differences are expected, what is the acceptable range of variation?
3. Could these differences impact more sensitive applications that rely on logit values rather than just the final classifications?
4. Are there known issues with XLM-RoBERTa models specifically in Candle?

**Comment by LaurentMazare** [comment]
Thanks for reporting this and for the detailed repro.
It seems that the discrepancy was caused by a bug on the candle side where the activation used in the classification head was a gelu whereas the original code used a tanh [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/xlm_roberta/modeling_xlm_roberta.py#L1454), I've put a fix in #2968.
Generally, you should expect values to be in line betwen pytorch and candle but with the caveats of numerical precision so when using a f32 you should have a bunch of matching digits (I would say at least 4 or 5), when using f16 like here a bit less, and when using bf16 even less matching digits. Here I get the following logits for f16 with the rust implementation which seems reasonably in line for f16.
```
Candle logits: [[2.6445313, -2.3457031], [1.0380859, -0.96191406], [1.4160156, -1.2705078], [3.4414063, -3.1738281]
```

**Comment by jpe90** [comment]
I've confirmed those logits are now coming through as expected when building off master, thank you so much for the explanation and quick fix @LaurentMazare.
